### PR TITLE
Migrate versioning to Gradle plugin and automate version bumps

### PR DESCRIPTION
**Pull request by msfjarvis**
## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [x] Refactoring
- [ ] Dependency updates

## :scroll: Description

1. Adds a `VersioningPlugin` that configures the versionCode and versionName properties of an application module and provides tasks to bump the version.
2. Applies the plugin to the 'app' module
3. Updates the Actions release drafter workflow to leverage the version modification capabilities provided by the plugin.

## :bulb: Motivation and Context

Makes releases much more easier to automate by removing another extra step.

## :green_heart: How did you test it?

- [Actions run](https://github.com/msfjarvis/Android-Password-Store/actions/runs/491386887)
- [Generated PR](https://github.com/msfjarvis/Android-Password-Store/pull/19)

## :pencil: Checklist

- [x] I formatted the code with the IDE's reformat action (Ctrl + Shift + L/Cmd + Shift + L)
- [x] I reviewed submitted code
- [ ] I added a [CHANGELOG](CHANGELOG.md) entry if applicable

## :crystal_ball: Next steps

~~Update release documentation to account for the change.~~ https://github.com/Android-Password-Store/docs/commit/e373de219f62

### CODE REVIEW

File: .github/workflows/draft_new_release.yml [hunk 1]
```diff
@@ -9,32 +9,72 @@ jobs:
...
-          echo "RELEASE_VERSION=${VERSION/v/}" >> $GITHUB_ENV
+          RELEASE_VERSION="${VERSION/v/}"
+          # Transforms 1.13.2 to 1.13 so that we can re-use the same
+          # branch for patch releases.
+          BRANCH_VERSION="${RELEASE_VERSION:$i:-2}"
+          if [[ "${RELEASE_VERSION: -1}" == "0" ]]; then
+            CHECKOUT_REF="develop"
+          else
+            CHECKOUT_REF="release-${BRANCH_VERSION}"
+          fi
```
Comment (msfjarvis) [inline .github/workflows/draft_new_release.yml:23]: When we release a new minor version (1.12.2 -> 1.13.0) we'll need to create the `release-1.13` branch first, so we checkout `develop` and later copy it into a `release-1.13` that gets pushed right before creating the PR to it. This block simply checks that the last digit is 0 indicating a new minor version.

File: .github/workflows/draft_new_release.yml [hunk 2]
```diff
@@ -9,32 +9,72 @@ jobs:
...
+
+          # Export variables separately so the scripting above is more legible,
+          # and we can actually use them within this block. Changes to $GITHUB_ENV
+          # only affect the next step, not the current one.
```
Comment (msfjarvis) [inline .github/workflows/draft_new_release.yml:27]: `$GITHUB_ENV` note is here to document what I learned today, no strong preference to keeping or removing it.

File: .github/workflows/draft_new_release.yml [hunk 3]
```diff
@@ -9,32 +9,72 @@ jobs:
...
+
+      - uses: actions/checkout@5a4ac9002d0be2fb38bd78e4b4dbde5606d7042f
+        with:
+          ref: ${{ env.CHECKOUT_REF }}
 
       - name: Update changelog
         uses: thomaseizinger/keep-a-changelog-new-release@96ebf19f2bddaf72406c84691b9c2d827da53140
         with:
-          version: ${{ env.RELEASE_VERSION }}
+          version: ${{ github.event.milestone.title }}
```
Comment (msfjarvis) [inline .github/workflows/draft_new_release.yml:41]: We need the 'v' prefix here for changelog links to correctly work (tags have the prefix) so I simply switched it to the milestone title which has the prefix.

File: .github/workflows/draft_new_release.yml [hunk 4]
```diff
@@ -9,32 +9,72 @@ jobs:
...
 
       - name: Initialize git config and commit changes
         run: |
+          # Configure name and email for Actions user
           git config user.name "GitHub Actions"
           git config user.email noreply@github.com
 
+          # It is necessary to create the $PR_BASE branch if it doesn't
+          # already exist because we want to start a PR against it.
```
Comment (msfjarvis) [inline .github/workflows/draft_new_release.yml:50]: As described above

File: app/version.properties [hunk 5]
```diff
@@ -0,0 +1,6 @@
+#
+#This file was automatically generated by 'versioning-plugin'. DO NOT EDIT MANUALLY.
+#
+#Sun Jan 17 12:32:03 IST 2021
```
Comment (msfjarvis) [inline app/version.properties:4]: `java.util.Properties` has this nasty 'feature' to unconditionally write this header. I would very much love to not have it.

File: buildSrc/buildDependencies.gradle [hunk 6]
```diff
@@ -3,11 +3,13 @@ rootProject.ext.versions = [
...
         download_plugin                : '4.1.1',
+        semver                         : '0.9.0',
 ]
 
 rootProject.ext.build = [
         androidGradlePlugin          : "com.android.tools.build:gradle:${versions.agp}",
         kotlinGradlePlugin           : "org.jetbrains.kotlin:kotlin-gradle-plugin:${versions.kotlin}",
         binaryCompatibilityValidator : "org.jetbrains.kotlinx:binary-compatibility-validator:${versions.binary_compatibility_validator}",
         downloadTaskPlugin           : "de.undercouch:gradle-download-task:${versions.download_plugin}",
+        jsemver                      : "com.github.zafarkhaja:java-semver:${versions.semver}",
```
Comment (msfjarvis) [inline buildSrc/buildDependencies.gradle:14]: [The library](https://github.com/zafarkhaja/jsemver) hasn't been touched in ~5 years but it does the job so :shrug:

File: buildSrc/src/main/java/VersioningPlugin.kt [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    /**
+     * Write an Android-specific variant of [this] to [stream]
+     */
+    private fun Version.writeForAndroid(stream: OutputStream) {
+        val newVersionCode = androidCode()
+        val props = Properties()
+        props.setProperty(VERSIONING_PROP_VERSION_CODE, "$newVersionCode")
+        props.setProperty(VERSIONING_PROP_VERSION_NAME, toString())
+        props.store(stream, VERSIONING_PROP_COMMENT)
```
Comment (msfjarvis) [inline buildSrc/src/main/java/VersioningPlugin.kt:50]: Updating this and the `versionProps` field below to use something that's not `java.util.Properties` is desirable due to the date header issue mentioned above.

Comment (fmeum) [reply buildSrc/src/main/java/VersioningPlugin.kt:50]: Maybe this is viable as it doesn't require much additional code: https://stackoverflow.com/a/60162749

Comment (msfjarvis) [reply buildSrc/src/main/java/VersioningPlugin.kt:50]: I have been to the page and like a perfectly meme-replicate developer, never scrolled past the accepted answer. Giving this a shot.

**Comment by msfjarvis** [comment]
Drafting to address a couple other bugs I found as well.

File: buildSrc/src/main/java/VersioningPlugin.kt (hunk 7 above)
Comment (msfjarvis) [reply buildSrc/src/main/java/VersioningPlugin.kt:50]: Yeah it didn't work and I don't care anymore. Let's ship.

**Review by Skrilltrax — APPROVED** [review]

**Pull request merged**